Integration tests for WALLETS_LIST

piotr-iohk opened the pull request:
<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

<p align="right">#</p>

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] Updated tests for WALLETS_LIST_01
- [x] I have added tests for WALLETS_LIST_02
- [x] I have added tests for WALLETS_LIST_03 (implemented but made partially pending due to #220)
- [x] I have added tests for WALLETS_LIST_04  (implemented but made pending due to #220)


# Comments
This can be addressed in subsequent PR or in this one, but:
 
- I'm missing method like `expectClientHttpError` method for asserting the `ClientHttpError` message -> addressed under #221.

- ~also to move forward I need to get some examples on how to use `Sort` and `Filter` in the safe/typed Client~
After thinking about it I think I'll go for `unsafeRequest` after all for sort and filter in order to have better code coverage... if I'm not mistaken the type-safe client would not reach all cases here -> https://github.com/input-output-hk/cardano-wallet/blob/d28b3f7b24ed3f163d46f8c18a91a117c7e5801e/src/Cardano/Wallet/API/Request/Sort.hs#L208 


<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

piotr-iohk commented on test/integration/Test/Integration/Scenario/Wallets.hs:
```diff
@@ -43,11 +43,146 @@ spec = do
+                        $- NoFilters
+                        $- NoSorts
+                    verify response expectations
+            in case (title == "1 wallet; page=9223372036854775807 & per_page=50 => 0 wallets returned") of
```
This test fails due to BUG https://github.com/input-output-hk/cardano-wallet/issues/213, hence made this scenario as pending...

piotr-iohk commented on test/integration/Test/Integration/Scenario/Wallets.hs:
```diff
@@ -43,11 +43,146 @@ spec = do
+            in case (title == "1 wallet; page=9223372036854775807 & per_page=50 => 0 wallets returned") of
+                True      -> xscenario title $ test
+                False     -> scenario title $ test
+
```
This particular test scenario fails due to BUG https://github.com/input-output-hk/cardano-wallet/issues/213, hence made it pending...

parsonsmatt replied:
You can mark a scenario pending with `pendingWith`. It may be necessary to write a lifted version of that function. But it should be:

```haskell
scenario title $ do
  pendingWith "This test fails due to bug #213"
  test
```

That way we get a note in the code that it is pending.

piotr-iohk commented on test/integration/Test/Integration/Scenario/Wallets.hs:
```diff
@@ -43,11 +43,146 @@ spec = do
+            setup $ defaultSetup
+                & walletName .~ show (name :: Int)
+                & initialCoins .~ [coins]
+
```
For now I made this one pending because setting up 3 wallets here was causing:
```
uncaught exception: IOException of type UserError
       user error (expectTxStatusEventually: waited too long for statuses.)
```

parsonsmatt replied:
Rather than `xscenario`, use `scenario`, and `pendingWith` with a note that says *why* it is failing.

parsonsmatt commented on test/integration/Test/Integration/Scenario/Wallets.hs:
```diff
@@ -43,11 +43,146 @@ spec = do
+                  )
+                ]
+
+        forM_ matrix $ \(title, wallets_number, page, per_page, expectations) ->
```
```suggestion
        forM_ matrix $ \(title, walletsNumber, page, perPage, expectations) ->
```

Haskell variables should be camel cased.

parsonsmatt commented on test/integration/Test/Integration/Scenario/Wallets.hs:
```diff
@@ -43,11 +43,146 @@ spec = do
+                        $- NoFilters
+                        $- NoSorts
+                    verify response expectations
+            in case (title == "1 wallet; page=9223372036854775807 & per_page=50 => 0 wallets returned") of
```
To remove ambiguity, this string should be factored out to a common value and reused rather than copy-pasted.

KtorZ commented on test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -347,6 +363,35 @@ walletName = typed
+    -> m ()
+expectDataListItemFieldEqual i getter a = \case
+    Left e  -> wantedSuccessButError e
+    Right s -> (view getter $ s!!(i - 1)) `shouldBe` a
```
Sweet :smile: !
Yet, this will blow up if the list is too small and also, elements are 0-indexed, it will be better to preserve that in the function arguments to not make it counter-intuitive.

Also, I'd make more re-use of the existing code and not re-write the getter machinery.

```suggestion
expectListItemFieldEqual 
    :: (MonadIO m, MonadFail m, Show a, Eq a)
    => Int
    -> Lens' s a
    -> a
    -> Either ClientError [s]
    -> m ()
expectListItemFieldEqual i getter a = \case
    Left _ -> expectFieldEqual getter a
    Right s 
      | length s > i -> expectFieldEqual getter a (Just (s !! i))
      | otherwise    -> fail $
        "expectListItemFieldEqual: trying to access the #" <> show i <> 
        " element from a list but there's none! Here's the list: " <> 
        show s
```

KtorZ commented on test/integration/Test/Integration/Scenario/Wallets.hs:
```diff
@@ -43,11 +43,133 @@ spec = do
+                    True  -> scenario title $ do
+                                pendingWith "Test fails due to bug #213"
+                                test
+                    False -> scenario title $ test
```
`pendingWith` actually "interrupts" the test scenario. So you could just avoid the `let ... in` binding an go with:

```
do
    forM_ ([1..walletsNumber]) $ \name -> do
        when (title == failingScenario) $
            pendingWith ...

        -- rest of the rest
```

piotr-iohk commented:
I have implemented requested changes and added new tests for WALLETS_LIST_03 and 04(Part of the tests is made pending due to #220), rebased. . Please review @KtorZ and/or @parsonsmatt. Thanks!

KtorZ commented on test/integration/Test/Integration/Scenario/Wallets.hs:
```diff
@@ -43,11 +42,464 @@ spec = do
+        let sortedWalletIds =
+                sort [ fromWalletId $ (fixtures !! 0) ^. wallet . walletId
+                     , fromWalletId $ (fixtures !! 1) ^. wallet . walletId
+                     ]
```
Hmm.. To be totally fair, I'd have sorted them "by hand". Especially because both tests and the actual implementation are written in Haskell and probably uses the same sorting function under the hood. Therefore, if our `Ord` instance was messed up, this test would catch it because it would just reproduce what's the backend is also doing. 

Yet, that's somewhat tough because we don't know the wallet ids upfront :upside_down_face: Ids are generated deterministically from the wallet secret key, so it's still possible to force some mnemonics and compute their corresponding wallet ids upfront.

KtorZ replied:
Well, `sort` is native yes, but it relies on the `Ord` instance of type you're trying to sort. 

```hs
sort :: Ord a => [a] -> [a]
```

Therefore:

```hs
data WalletId = WalletId String deriving (Show, Eq)

instance Ord WalletId where
  (WalletId a) <= (WalletId b) = a >= b

main :: IO ()
main = do
  print $ sort $ [WalletId "a", WalletId "b", WalletId "c"]
  -- [WalletId "c",WalletId "b",WalletId "a"]
```

I suppose a comment would be enough here since we're _probably_ (I hope) using the stock `Ord` instance for the `WalletId` and that we aren't defining it by hand.

piotr-iohk replied:
Ok, got it.
However I was not sorting `WalletId` but actually a `Text` since: `fromWalletId :: Client.WalletId -> Text`.

Maybe the name of the list should have been more descriptive then, like: `sortedWalletIdsText` instead of `sortedWalletIds`...

KtorZ commented on test/integration/Test/Integration/Scenario/Wallets.hs:
```diff
@@ -43,11 +42,464 @@ spec = do
+        let sortedWalletIds =
+                sort [ fromWalletId $ (fixtures !! 0) ^. wallet . walletId
+                     , fromWalletId $ (fixtures !! 1) ^. wallet . walletId
+                     ]
```
Same remark as before :)

KtorZ commented on test/integration/Test/Integration/Scenario/Wallets.hs:
```diff
@@ -43,11 +42,464 @@ spec = do
+        let sortedWalletIds = sort walletIds
+
+        let endpoint = "api/v1/wallets?sort_by=created_at&id=RANGE%5B" <+> ( sortedWalletIds !! 0 )
+                        <+> "," <+> ( sortedWalletIds !! 2 )  <+> ("%5D" :: Text)
```
Why sorting the wallets by ids here? The sort given in the query is by `created_at` ?

piotr-iohk replied:
Yes. Sorting the ids because I want to use `RANGE[smallest, biggest]` and then I sort the results by `created_at` to make sure all the expected wallet id's are present in the returned results (because I know they will be presented in the descending order)...

KtorZ commented on test/integration/Test/Integration/Scenario/Wallets.hs:
```diff
@@ -405,6 +856,10 @@ spec = do
   where
+
+    fromWalletId :: Client.WalletId -> Text
+    fromWalletId (Client.WalletId a) = a
```
Looks like we could use a `Lens` here to unwrap things like this. This would play nicely with `fixture ^. wallet . walletId . asText` 

(not asking you to implement that, more like leaving a note for myself :p )

piotr-iohk commented on test/integration/Test/Integration/Scenario/Wallets.hs:
```diff
@@ -43,11 +42,464 @@ spec = do
+
+        forM_ matrix $ \(endpoint, expectations) -> scenario endpoint $ do
+            pendingWith "Test fails due to bug #220"
+            forM_ ([3,6,6,9,0]) $ \coins -> do
```
OK, however, I have experienced the issue from #220 (also pretty much repeatedly) when I had no `0` coin with something like:
```
            forM_ (zip [1,2,3] [3,2,1]) $ \(name, coins) -> do
                setup $ defaultSetup
                    & walletName .~ show (name :: Int)
                    & initialCoins .~ [coins]
```

piotr-iohk commented on test/integration/Test/Integration/Scenario/Wallets.hs:
```diff
@@ -43,11 +42,464 @@ spec = do
+                [ (fixtures !! 0) ^. wallet . walletId
+                , (fixtures !! 1) ^. wallet . walletId
+                ]
+        let endpoint = "api/v1/wallets?id=" <+> ( fromWalletId $ walletIds !! 0 )
```
>Why are you actually defining those multiple walletIds actually ^^?

Well, that's part of the test actually. Let's assume that filtering by id stopped working and returns all wallet's even if I specify particular one. If I have only one wallet in this test, it will not catch that :)
